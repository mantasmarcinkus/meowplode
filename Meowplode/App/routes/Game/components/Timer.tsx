import * as React from "react";
import { ActionCreators, DefinedState } from "../module";

const FONT_SIZE = 60;
const COLOR = "#FF530D";

interface Props extends Pick<ActionCreators, "timeOut"> {
  secondsRemaining: number;
}

interface State {
  secondsRemaining: number;
  fontSize: number;
  color: string;
}
export default class Timer extends React.Component<Props, State> {
  private interval: any; // Typescript and javascript hacking is great. Using the same variable for different types.
  constructor(props: Props) {
    super(props);
    this.state = {
      color: COLOR,
      fontSize: FONT_SIZE,
      secondsRemaining: props.secondsRemaining,
    };
    this.tick = this.tick.bind(this);
  }
  public tick() {
    this.setState({
      color: this.getColor(this.state.secondsRemaining),
      fontSize: this.getFontSize(this.state.fontSize, this.state.secondsRemaining),
      secondsRemaining: this.state.secondsRemaining - 1,
    });

    if (this.state.secondsRemaining <= 0) {
      this.props.timeOut();
      clearInterval(this.interval);
    }
  }
  public componentDidMount() {
    this.setState({ secondsRemaining: this.props.secondsRemaining, fontSize: FONT_SIZE, color: COLOR });
    this.interval = setInterval(this.tick, 1000);
  }
  public componentWillUnmount() {
    clearInterval(this.interval);
  }
  public render() {
    const minutes = ("0" + parseInt((this.state.secondsRemaining / 60).toString(), 10)).slice(-2);
    const seconds = ("0" + parseInt((this.state.secondsRemaining % 60).toString(), 10)).slice(-2);
    return (
      <div className="digital" style={{ fontSize: this.state.fontSize, color: this.state.color, height: 100 }}>
        <span>{minutes}:{seconds}</span>
      </div>
    );
  }
  private getFontSize(fontSize: number, secondsRemaining: number) {
    return secondsRemaining < 7 ? FONT_SIZE * 1.3 : FONT_SIZE;
  }
  private getColor(secondsRemaining: number) {
    return secondsRemaining < 7 ? "red" : COLOR;
  }
}
