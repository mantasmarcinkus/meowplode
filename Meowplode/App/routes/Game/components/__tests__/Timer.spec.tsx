import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import Timer from "../Timer";

jest.mock("../../../Leaderboard/container");

const testProps: any = {
  secondsRemaining: 10,
};

describe("<Timer />", () => {
  let wrapper: ShallowWrapper<any, any>;
  beforeEach(() => {
    wrapper = shallow(<Timer {...testProps} />);
  });
  it("renders", () => {
    expect(wrapper.render()).toMatchSnapshot();
  });
  it("state is in good shape", () => {
    expect(wrapper.instance().state.secondsRemaining).toEqual(10);
    expect(wrapper.instance().state.color).toEqual("#FF530D");
    expect(wrapper.instance().state.fontSize).toEqual(60);
  });
  it("state tick updates state", () => {
    (wrapper.instance() as any).tick();
    expect(wrapper.instance().state.secondsRemaining).toEqual(9);
  });
});
