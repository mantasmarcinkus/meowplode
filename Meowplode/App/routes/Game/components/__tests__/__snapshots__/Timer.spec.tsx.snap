// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> renders 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "digital",
          "style": "font-size:60px;color:#FF530D;height:100px;",
        },
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "data": "00:10",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "span",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
          },
        ],
        "name": "div",
        "next": null,
        "parent": null,
        "prev": null,
        "root": [Circular],
        "type": "tag",
      },
    ],
    "name": "root",
    "next": null,
    "parent": null,
    "prev": null,
    "type": "root",
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xmlMode": false,
  },
}
`;
